bee-network/bee-autopeering/src/peering/neighbor.rs

```diff
@@ -0,0 +1,323 @@
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Ord for Neighbor {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.distance.cmp(&other.distance)
+    }
+}
+
+impl AsRef<Peer> for Neighbor {
+    fn as_ref(&self) -> &Peer {
+        &self.peer
+    }
+}
+
+#[derive(Clone, Default)]
+pub(crate) struct Neighborhood<const N: usize, const INBOUND: bool> {
+    inner: Arc<RwLock<NeighborhoodInner<N, INBOUND>>>,
+}
+
+impl<const N: usize, const INBOUND: bool> Neighborhood<N, INBOUND> {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn lock_select(&self, candidate: Neighbor) -> Option<Peer> {
+        let mut write = self.write();
+        write.select(candidate)
+    }
+
+    // TODO: do not expose lock guards, because this leads to dead locks if not very careful.
+    pub fn read(&self) -> RwLockReadGuard<NeighborhoodInner<N, INBOUND>> {
+        self.inner.read().expect("error getting read access")
+    }
+
+    // TODO: do not expose lock guards, because this leads to dead locks if not very careful.
+    pub fn write(&self) -> RwLockWriteGuard<NeighborhoodInner<N, INBOUND>> {
+        self.inner.write().expect("error getting write access")
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NeighborhoodInner<const N: usize, const INBOUND: bool> {
+    neighbors: Vec<Neighbor>,
+}
+
+impl<const N: usize, const INBOUND: bool> NeighborhoodInner<N, INBOUND> {
+    // TODO: revisit dead code
+    #[allow(dead_code)]
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn insert_neighbor(&mut self, peer: Peer, local: &Local) -> bool {
+        // If the peer already exists remove it.
+        // NOTE: It's a bit less efficient doing it like this, but the code requires less branching this way.
+        let _ = self.remove_neighbor(peer.peer_id());
+
+        if self.neighbors.len() >= N {
+            return false;
+        }
+
+        // Calculate the distance to that peer.
+        let distance = salt_distance(local.read().peer_id(), peer.peer_id(), &{
+            if INBOUND {
+                local.read().private_salt().expect("missing private salt").clone()
+            } else {
+                local.read().public_salt().expect("missing public salt").clone()
+            }
+        });
+
+        self.neighbors.push(Neighbor { distance, peer });
+
+        true
+    }
+
+    pub(crate) fn remove_neighbor(&mut self, peer_id: &PeerId) -> Option<Peer> {
+        if self.neighbors.is_empty() {
+            None
+        } else if let Some(index) = self.neighbors.iter().position(|pd| pd.peer().peer_id() == peer_id) {
+            let Neighbor { peer, .. } = self.neighbors.remove(index);
+            Some(peer)
+        } else {
+            None
+        }
+    }
+
+    /// Check whether the candidate is a suitable neighbor.
+    pub(crate) fn select(&mut self, candidate: Neighbor) -> Option<Peer> {
```

Comment:
This is a rather weird API tbh, why does it take ownership of the Neighbor ? Why does it return an Option and not a boolean ?

Reply:
You're right. The API is not good. Fixing...

Reply:
Fixed.